ecommerce/models.py

```diff
@@ -154,3 +161,116 @@ def save(self, *args, **kwargs):
     def __str__(self):
         """Description for CoursePrice"""
         return "CoursePrice for {}, price={}, is_valid={}".format(self.course_run, self.price, self.is_valid)
+
+
+class Coupon(Model):
+    """
+    Model for a coupon. This stores the discount for the coupon, how many people can use it, and how many times
+    a person can use it.
+
+    When a coupon is redeemed by a purchaser the counter on this object is decremented
+    and a UserCoupon object is created for that particular purchaser.
+    """
+    PERCENT_DISCOUNT = 'percent-discount'
+    FIXED_DISCOUNT = 'fixed-discount'
+
+    AMOUNT_TYPES = [PERCENT_DISCOUNT, FIXED_DISCOUNT]
+
+    coupon_code = TextField(
+        null=True,
+        blank=True,
+        help_text="""The coupon code used for redemption by the purchaser in the user interface.
+    If blank, the purchaser may not redeem this coupon through the user interface,
+    though it may be redeemed in their name by an administrator.""",
+    )
+    content_type = ForeignKey(
+        ContentType,
+        on_delete=SET_NULL,
+        null=True,
+        help_text="content_object is a link to either a Course, CourseRun, or a Program",
+    )
+    object_id = PositiveIntegerField()
+    content_object = GenericForeignKey('content_type', 'object_id')
+
+    amount_type = CharField(
+        choices=[(_type, _type) for _type in AMOUNT_TYPES],
+        max_length=30,
+        help_text="Whether amount is a percent or fixed discount",
+    )
+    amount = DecimalField(
+        decimal_places=2,
+        max_digits=20,
+        help_text="Either a number from 0 to 1 representing a percent, or the fixed value for discount",
+    )
+
+    num_coupons_available = PositiveIntegerField(
+        null=False,
+        help_text="Number of people this coupon can be redeemed by",
+    )
+    num_redemptions_per_user = PositiveIntegerField(
+        null=False,
+        help_text="Number of times a person can redeem a coupon",
+    )
+    expiration_date = DateTimeField(
+        null=True,
+        blank=True,
+        help_text="If set, the coupons will not be redeemable after this time",
+    )
+    activation_date = DateTimeField(
+        null=True,
+        blank=True,
+        help_text="If set, the coupons will not be redeemable before this time",
+    )
+    enabled = BooleanField(default=True, help_text="If true, coupons are presently redeemable")
```

Comment:
I would suggest adding a `is_redeemable` utility method to this class:

```python
def is_redeemable(self):
    if not self.enabled:
        return False
    start = self.activation_date or datetime.min
    end = self.expiration_date or datetime.max
    return start <= datetime.now() <= end
```

It encapsulates checking all the conditions that could cause this coupon to not be redeemable at this point in time. If we add other possible conditions in the future, those could be rolled into this method as well.